pkg/sql/sqlutil/internal_executor.go

```diff
@@ -196,16 +207,32 @@ type InternalRows interface {
 	Types() colinfo.ResultColumns
 }
 
-// SessionBoundInternalExecutorFactory is a function that produces a "session
-// bound" internal executor.
-type SessionBoundInternalExecutorFactory func(
+// InternalExecutorFactory is a function that produces an internal executor.
+type InternalExecutorFactory func(
```

Comment:
I'm not sure, seems to me that changing it to an interface will create circular dependency headaches. 

Currently we define the factories in `server/server_sql.go`, which means we can call `sql.MakeInternalExecutor()` at ease. But if it's changed to an interface, to define the function under structs (such as `Registry`), don't we have to import the dependencies into their packages?

Reply:
I don't follow. The implementation of the interface can be in `sql`, the interface can be in `sqlutil`.

Reply:
Maybe I misunderstood, but I thought you wanted to have something like
```go
type SessionBoundInternalExecutorFactory interface {
InternalExecutorFactory(context.Context, *sessiondata.SessionData) InternalExecutor 
}
```

And any struct that currently has a `SessionBoundInternalExecutorFactory` field should now implement the `InternalExecutorFactory()` method. 

So the question is, for structs that are not in the `sql` package, it may be hard to implement `InternalExecutorFactory()` in their packages. For example, the implementation of `Registry.InternalExecutorFactory()` has to happen under `jobs` (because IIUC everything related to one struct must be defined in the same package), but it has to call `sql.MakeInternalExecutor()`, which will bring circular dependency. 

Sorry if I got the wrong idea, but currently I can't think of any other interpretation of "change this to be an interface".

Reply:
>And any struct that currently has a SessionBoundInternalExecutorFactory field should now implement the InternalExecutorFactory() method.

Rather, I'm suggesting that every struct which today has a SessionBoundInternalExecutorFactory would instead have an `InternalExecutorFactory` field which has type `InternalExecutorFactory` where that type is an interface and not a function. That interface may have a method, say, `NewInternalExecutor` which, in the first pass, does what the function today does.

Reply:
Thanks for the explanation!

I feel I understand most of it now, but I'm still a bit unsure how to define the `InternalExecutorFactory` field of structs. So I made these tentative changes: https://gist.github.com/ZhouXing19/f4a7c11a0e4b9d3d440cc10007707c86. Is this what you had in mind?

Reply:
Yeah, that's what I had in mind except that the fields of the implementation of the `InternalExecutorFactory` shouldn't be exported and there should be a constructor.

Reply:
Thanks! Made the changes.